File: src/process/flip/flip.js
/**
 * Flip process step.
 *
 * Mirrors the main geometry around the XY plane at Z=0, normalizes the flipped
 * geometry so zMin becomes 0, and updates cursorZ to the highest direct root
 * body top.
 */
export function execute({ state, geometryState }) {
  const mainState = requiredGeometryState(
    geometryState?.("main_geometry") ?? state,
    "main_geometry",
  );

  mainState.flipAroundZ({
    z: 0,
    normalizeZMinToZero: true,
    updateCursor: false,
  });
  mainState.setCursorZ(mainState.rootBodyZMax());

  return mainState;
}

function requiredGeometryState(value, label) {
  if (
    !value ||
    typeof value.flipAroundZ !== "function" ||
    typeof value.rootBodyZMax !== "function" ||
    typeof value.setCursorZ !== "function"
  ) {
    throw new Error(`Flip.${label} must resolve to a ProcessGeometryState`);
  }
  return value;
}

File: packages/process-step-py/src/process_flow_steps/flip/flip.py
from process_flow_kernel import ProcessGeometryState, ProcessStepContext


def execute(context: ProcessStepContext) -> ProcessGeometryState:
    state = context.state
    state.flip_around_z(z=0, normalize_z_min_to_zero=True, update_cursor=False)
    state.set_cursor_z(state.root_body_z_max())
    return state
